src/rate/index.tsx

```diff
@@ -0,0 +1,336 @@
+	max?: number;
+	/** Allow undefined to be chosen with repeat click or with arrows */
+	allowClear?: boolean;
+	/** Prevent interaction with a visual indication */
+	disabled?: boolean;
+	/** Prevent interaction */
+	readOnly?: boolean;
+}
+
+export interface RateChildren {
+	(fill: boolean, integer: number, selected?: number, over?: number): RenderResult;
+}
+
+interface RateState {
+	uuid: string;
+	hover: string;
+	hovering: boolean;
+	selectedInteger?: number;
+	selectedStep?: number;
+	focused: boolean;
+}
+
+const icache = createICacheMiddleware<RateState>();
+
+function mixedNumber(integer: number, step: number, steps: number): MixedNumber {
+	if (step === steps) {
+		step = 0;
+	} else {
+		integer -= 1;
+	}
+	return {
+		value: integer + step / steps,
+		quotient: integer,
+		numerator: step,
+		denominator: steps
+	};
+}
+
+type RadioChecked = (checked?: boolean) => boolean | undefined;
+
+const factory = create({ focus, i18n, icache, theme })
+	.properties<RateProperties>()
+	.children<RateChildren | undefined>();
+
+export const Rate = factory(function Rate({
+	id,
+	children,
+	properties,
+	middleware: { focus, i18n, icache, theme }
+}) {
+	const { format } = i18n.localize(bundle);
+	const themedCss = theme.classes(css);
+	const {
+		name,
+		label,
+		initialValue: initialNumber = undefined,
+		onValue,
+		max = 5,
+		steps: stepsLength = 1,
+		allowClear = false,
+		disabled = false,
+		readOnly = false
+	} = properties();
+	const interaction = !disabled && !readOnly;
+	const initialStep = initialNumber ? Math.round((initialNumber % 1) * stepsLength) : undefined;
+	const initialInteger = initialNumber ? Math.ceil(initialNumber) : undefined;
+	const initialValue = initialNumber
+		? `${initialInteger}-${initialStep || stepsLength}`
+		: undefined;
+	const character = children()[0];
+	const hovering = icache.getOrSet('hovering', false);
+	const selectedInteger = icache.getOrSet('selectedInteger', initialInteger);
+	const selectedStep = icache.getOrSet('selectedStep', initialStep);
+	const shouldFocus = hovering ? focus.shouldFocus() : icache.getOrSet('focused', false);
+
+	const integers: number[] = [];
+	const steps: number[] = [];
+	for (let step = 1; step <= stepsLength; step++) {
+		steps.push(step);
+	}
+	const options = [];
+	if (allowClear) {
+		integers.push(0);
+		options.push({
+			value: '',
+			label: format('starLabels', { quotient: 0, numerator: 0, denominator: 1 })
+		});
+	}
+	for (let integer = 1; integer <= max; integer++) {
+		integers.push(integer);
+		for (const step of steps) {
+			options.push({
+				value: `${integer}-${step}`,
+				label: format('starLabels', mixedNumber(integer, step, stepsLength))
+			});
+		}
+	}
+
+	const _onValue = (value?: string) => {
+		if (!value) {
+			icache.set('selectedInteger', undefined);
+			icache.set('selectedStep', undefined);
+			onValue && onValue();
+		} else {
+			const [integer, step] = value.split('-').map((num) => +num);
```

Comment:
I think that this would be a lot more straight forward if decimal values were simply passed to the radio group as options as you could just parse the radio group's `onValue` callback into an int and return it.